tlv/record.go

```diff
@@ -0,0 +1,147 @@
+package tlv
+
+import (
+	"io"
+
+	"github.com/btcsuite/btcd/btcec"
+)
+
+// Type is an 64-bit identifier for a TLV Record.
+type Type uint64
+
+// Encoder is a signature for methods that can encode TLV values. An error
+// should be returned if the Encoder cannot support the underlying type of val.
+// The provided scratch buffer must be non-nil.
+type Encoder = func(w io.Writer, val interface{}, buf *[8]byte) error
+
+// Decoder is a signature for methods that can decode TLV values. An error
+// should be returned if the Decoder cannot support the underlying type of val.
+// The provided scratch buffer must be non-nil.
+type Decoder = func(r io.Reader, val interface{}, buf *[8]byte, l uint64) error
+
+// SizeFunc is a function that can compute the length of a given field. Since
+// the size of the underlying field can change, this allows the size of the
+// field to be evaluated at the time of encoding.
+type SizeFunc func() uint16
+
+// SizeVarBytes is a SizeFunc that can compute the length of a byte slice.
+func SizeVarBytes(e *[]byte) SizeFunc {
+	return func() uint16 {
+		return uint16(len(*e))
+	}
+}
+
+// Record holds the required information to encode or decode a TLV record.
+type Record struct {
```

Comment:
I don't think it matters if the padding is at the middle or at the end, but `structlayout` recommends a slightly different ordering: 
```
⛰   structlayout github.com/lightningnetwork/lnd/tlv Record
Record.typ github.com/lightningnetwork/lnd/tlv.Type: 0-8 (size 8, align 8)
Record.staticSize uint16: 8-10 (size 2, align 2)
padding: 10-16 (size 6, align 0)
Record.value interface{}: 16-32 (size 16, align 8)
Record.sizeFunc github.com/lightningnetwork/lnd/tlv.SizeFunc: 32-40 (size 8, align 8)
Record.encoder func(w io.Writer, val interface{}, buf *[8]byte) error: 40-48 (size 8, align 8)
Record.decoder func(r io.Reader, val interface{}, buf *[8]byte, l uint64) error: 48-56 (size 8, align 8)
```

```
⛰   structlayout -json github.com/lightningnetwork/lnd/tlv Record | structlayout-optimize
Record.value interface{}: 0-16 (size 16, align 8)
Record.typ github.com/lightningnetwork/lnd/tlv.Type: 16-24 (size 8, align 8)
Record.sizeFunc github.com/lightningnetwork/lnd/tlv.SizeFunc: 24-32 (size 8, align 8)
Record.encoder func(w io.Writer, val interface{}, buf *[8]byte) error: 32-40 (size 8, align 8)
Record.decoder func(r io.Reader, val interface{}, buf *[8]byte, l uint64) error: 40-48 (size 8, align 8)
Record.staticSize uint16: 48-50 (size 2, align 8)
padding: 50-56 (size 6, align 0)
```

Reply:
good suggestion, done!